### ISSUE TITLE
Investigate Keda for event driven autoscaling

### ISSUE THREAD

**Issue by StevenBarre** [BCDevOps/developer-experience#3127]
**Describe the issue**
https://keda.sh/ is an operator that allows scaling HPAs to zero when there's no work, and then scaling it up when there is work in a work queue like Redis.

**What is the Value/Impact?**
Allow teams to respond to work queues and scale to zero when not needed.

**What is the plan? How will this get completed?**
Test out in a lab. Make sure it is multitenant safe.

**Identify any dependencies**

**Definition of done**
A decision on if we will roll this out into prod clusters and a plan to do so.

**Comment by sookeke** [comment]
[keda](https://github.com/kedacore/keda) will allow us to build Event-based architecture systems by extending the default feature of Kubernetes into serverless event-driven containers. it allows consumers of events/messages to handle workloads only when an event/message has been published to an event source (Kafka, azure service bus, rabbitMQ, Redis etc)

ISB needed this operator for Domain-Driven Event-based systems like the digital evidence management system.

**Comment by ChrisHoban** [comment]
As an Architect in the sector, I second this proposal: This will allow us to use even less platform resources with better performance for when it is actually needed. It will also enable better implementation of event driven patterns, wherein consumers of events can be implemented serverless. There are a number of existing projects on the platform that could substantially reduce resource footprint, both for regular services (Portals/APIs) and several idle "Event Consumer Services" that unfortunately get deployed to a desired watermark of event throughput (IE multiple pods/multiple threads sitting idle).

**Comment by sookeke** [comment]
@StevenBarre  Do we have any update for this? Still waiting to utilize this service as a top priority for our project. Thanks

**Comment by mitovskaol** [comment]
@StevenBarre I support this priority for including in the Sprint Backlog for the upcoming Sprint (Oct 26-Nov 15) for the Ops team , lets install this operator in KLAB first and test it there.

@ChrisHoban @sookeke  I created a project set for you in KLAB with the license plate `e5b08a` and gave you admin access to it. Once Steven installs the operation, he will let you know and you can test it in there.

**Comment by sookeke** [comment]
@mitovskaol  Thank you. I can deploy some of our Event-driven data mesh workload there. Let me know when the KEDA operator is installed. Do not forget to install both [Core](https://github.com/kedacore/keda) and [HTTP Add-on](https://github.com/kedacore/http-add-on) operator.

**Comment by StevenBarre** [comment]
> Community Operators are Operators which have not been vetted or verified by Red Hat. Community Operators should be used with caution because their stability is unknown. Red Hat provides no support for community Operators.

Just being clear that there is no support for this operator if something breaks.

**Comment by StevenBarre** [comment]
![image.png](https://images.zenhubusercontent.com/5d544a7a5faedc627998cd85/02a0d883-28dc-48d8-ad0f-43bd595bcffb)

The HTTP Add-on doesn't seem ready for production.

**Comment by StevenBarre** [comment]
https://github.com/kedacore/keda-olm-operator/issues/162

**Comment by StevenBarre** [comment]
Keda core installed into KLAB. Set up a ScaledObject in the `stevenbarre-test` namespace and had it query a Postgresql DB for its scale value. Got a deployment to scale to zero and scale up based on the value in the DB.

@sookeke @ChrisHoban have access to the `e5b08a` namespaces in KLAB for testing and have said there is enough quota. Moving ticket to Blocked while they do their testing and report back.

**Comment by StevenBarre** [comment]
@ChrisHoban @sookeke have you had a chance to do any testing?

**Comment by sookeke** [comment]
I started deployment today I will let you know how things are going in next couple days @StevenBarre

**Comment by sookeke** [comment]
Hi Folks - 

I just concluded testing for KEDA core and scaled event using one of our openshift workloads.

**Observations**

1.  Keda has the ability to scale Kubernetes deployment to zero while listening to an external event like Kafka topic
2. Keda acts as a consumer to this topic by subscribing to the topic via scaledobject
3.  Keda does not support **Oauth** for SASL mechanism. It supports plaintext, scram_sha256 or scram_sha512 
4. Easy to use and integrates but will require developers with an intermediate understanding of DDD and Event-based architecture for distributed system
5. Has the ability to save CPU cost as this will make the Kubernetes infrastructure serverless

Can we move to install this operator in SIlver and GOLD?.

**Comment by sookeke** [comment]
> 

Testing is complete for kafka topics

**Comment by sookeke** [comment]
@StevenBarre  Did you delete the operator for KLAB?

**Comment by StevenBarre** [comment]
Yes, testing our automated install tool, and also giving the new RH version a test https://cloud.redhat.com/blog/custom-metrics-autoscaler-on-openshift

**Comment by sookeke** [comment]
@StevenBarre when do you think we should have this in silver or gold i.e CMA?

**Comment by StevenBarre** [comment]
Unsure when it will be ready. I've emailed Olena about the risks of installing this in Production clusters to seek her approval before moving forward.

**Comment by mitovskaol** [comment]
@StevenBarre  I confirmed with @sookeke that the AG team is ready to accept the consequences of using the operator that is still in Tech Preview. as long as the risk of the negative impact on the cluster or other team's app is low, I approve its roll out in Silver and Gold.

**Comment by StevenBarre** [comment]
OK. Code is merged to the CCM. Will get rolled out to Silver next Tues, Gold on Wednesday.

As the NSX clusters are still 4.9 they do not support the operator and won't be getting it until they are upgraded to 4.10 at some point.

**Comment by sookeke** [comment]
@StevenBarre  We moved all of our Kafka clusters to use SASL/OAuth as an authentication mechanism. That means the KEDA scaler needs to support OAuth. I realized that the operator version installed is v2.7.1. This version currently does not support OAuth. We will KEDA version greater than [KEDA 2.9.0](https://github.com/kedacore/keda/releases/tag/v2.9.0) to be able to connect to our clusters

**Comment by sookeke** [comment]
Probably need to re-open this issue

**Comment by StevenBarre** [comment]
Case opened with Red Hat asking for an ETA on when the operator will update its KEDA version https://access.redhat.com/support/cases/#/case/03422451

**Comment by sookeke** [comment]
@StevenBarre Thank you

**Comment by StevenBarre** [comment]
RH says the operator will likely update to 2.9 sometime in the summer, likely alongside the release of OCP 4.13

**Comment by StevenBarre** [comment]
@sookeke FYI there's some discussion ongoing about moving Emerald to the EUS versions of OCP. This will further delay that clusters access to the version of Keda with SASL/OAuth.
